doc/md/_a3_Getting-Started-Audio.md

```diff
@@ -0,0 +1,253 @@
+
+### Audio Data Packing
+
+On the daisy, there are two ways the audio data can be prepeared, "Interleaved" and "Non-Interleaved", the overall callback structure functions the same in either case, but the way the input and output data is packed is different in each one.
+
+Above, we looked at the non-interleaved format, where each channel had its own array of samples.
+
+Interleaved audio is instead one large array containing multiple channels of data. This data is interleaved so that there is a single sample of each channel before moving to the next sample. For example, with a block size of four, the full buffer would look like:
+
+```
+{ L0, R0, L1, R1, L2, R2, L3, R3 }
+```
+
+as opposed to the non-interleaved version of the same data, that could be represented as:
+
+```
+{ 
+    { L0, L1, L2, L3 },
+    { R0, R1, R2, R3 }
+}
+```
+
+## Changing Audio Settings
+
+The AudioHandle allows the user to change certain settings, in particular the:
+
+* Sample Rate
+* Block Size
+
+On the DaisySeed, or other official boards you can use the following functions:
+```cpp
+hw.SetAudioBlockSize(4);
+hw.SetAudioSampleRate(SaiHandle::Config::SampleRate::SAI_48KHZ);
+```
+
+Block size can be any number up to 256.
+
+See the [SaiHandle::Config::SampleRate](@ref daisy::SaiHandle::Config::SampleRate) enum for the available samplerate options. It's also worth mentioning that, depending on what codec is connected, not all options are compatible.
+
+## What Not to do in the Audio Callback
+
+The audio callback is in what's called and "Interrupt Handler" without going in depth on what that means here, it does mean that it the function _must_ take less time than it takes to transfer out the audio buffer, otherwise your audio will start to have under-run errors, which cause digital artifacts in the audio path.
+
+The amount of time can be calculated as `(1 / sample_rate) * block_size`. For example, with 48 sample block size at 48kHz, the function can take 1ms before underruns start to occur.
+
+Now, that is a bit of a simplification, because chances are you'll be doing a few things in the main `while()` loop, but it does mean that you shouldn't do things that take an indeterminant amount of time.
+
+These things include:
+
+* calls to other peripheral blocking functions (i.e. DAC Write, SPI tranfer, etc.)
+* dynamic allocation like `malloc`
+
+If you do need to trigger non-deterministic events from within the audio callback you can use a flag, or other mechanism that can be checked in the main loop. An overly simplified example of this might look like:
+
+```cpp
+// Global:
+bool action_flag;
+
+// In callback:
+if (some_event) // like a button press, or something
+    action_flag = true; 
+
+// In main()
+
+while(1) 
+{
+    
+    if (action_flag) {
+        // Do the big thing that can take a while
+        initiate_big_transfer();
+        // and clear the flag
+        action_flag = false;
+    }
+}
+```
```

Comment:
This could be added:

## How to measure the processing load

It is crucial to have a feeling of how much processing load your algorithm introduces and how much room for additions you still have. This can be done by measuring the execution time of the audio callback and comparing it to the available time. Daisy comes with a helper class to do this, the `CpuLoadMeter`.

To measure the CPU load, you would call `.OnBlockStart()` at the beginning, and `.OnBlockEnd()` at the end of your audio callback. The you can request the resulting load measurement and print it on a display or a serial connection. It is crucial to do the printing from the main loop, not the audio callback itself, as discussed before.

A complete example would look like this:

```cpp
#include "daisy_seed.h"
#include "daisysp.h"
using namespace daisy;
DaisySeed hw;
CpuLoadMeter loadMeter;

void MyCallback(AudioHandle::InputBuffer in, 
                AudioHandle::OutputBuffer out, 
                size_t size) 
{
    loadMeter.OnBlockStart();
    for (size_t i = 0; i < size; i++)
    {
        // add your processing here
        out[0][i] = 0.0f;
        out[1][i] = 0.0f;
    }
    loadMeter.OnBlockEnd();
}
int main(void)
{
    hw.Init();
    
    // start logging to the serial connection
    hw.StartLog();
    
    // initialize the load meter so that it knows what time is available for the processing:
    loadMeter.Init(hw.AudioSampleRate(), hw.AudioBlockSize());
    
    // start the audio processing callback
    hw.StartAudio(MyCallback);
    
    while(1) {
        // get the current load (smoothed value and peak values)
        const float avgLoad = cpuLoadMeter.GetAvgCpuLoad();
        const float maxLoad = cpuLoadMeter.GetMaxCpuLoad();
        const float minLoad = cpuLoadMeter.GetMinCpuLoad();
        // print it to the serial connection (as percentages)
        hw.PrintLine("Processing Load %:");
        hw.PrintLine("Max: " FLT_FMT3, FLT_VAR3(maxLoad * 100.0f));
        hw.PrintLine("Avg: " FLT_FMT3, FLT_VAR3(avgLoad * 100.0f));
        hw.PrintLine("Min: " FLT_FMT3, FLT_VAR3(minLoad * 100.0f));
        // don't spam the serial connection too much
        System::Delay(500);
    }
}
```

Reply:
Excellent suggestion! Added.